From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 9 Oct 2019 04:38:54 -0700
Subject: [PATCH] Lag compensate eating

When the server is lagging, players will wait longer when eating.
Change to also use a time check instead if it passes.

This can be controlled by config, and is enabled by default

diff --git a/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java b/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
index 6bc4c213dc9fc37e15f7506790ff5eb3e5294109..2969a219d6171f97b8407e42382d0eb34cf92c7a 100644
--- a/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
+++ b/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
@@ -115,6 +115,12 @@ public final class TuinityConfig {
         }
     }
 
+    public static boolean lagCompensateEating;
+
+    private static void lagCompensateEating() {
+        lagCompensateEating = TuinityConfig.getBoolean("lag-compensate-eating", true);
+    }
+    
     public static final class WorldConnfig {
 
         public final String worldName;
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 6184cced7319a6235db040dee205c50ef4c1ee74..2885ffe162fdbc9137c03552bf094510f32e8801 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -113,7 +113,7 @@ public abstract class EntityLiving extends Entity {
     private int jumpTicks;
     private float bD;
     public ItemStack activeItem; // Paper - public
-    protected int bl;
+    protected int bl; protected final int getEatTimeTicks() { return this.bl; } protected final void setEatTimeTicks(int value) { this.bl = value; } // Tuinity - OBFHELPER
     protected int bm;
     private BlockPosition bE;
     private DamageSource bF;
@@ -2844,6 +2844,8 @@ public abstract class EntityLiving extends Entity {
         return ((Byte) this.datawatcher.get(EntityLiving.ao) & 2) > 0 ? EnumHand.OFF_HAND : EnumHand.MAIN_HAND;
     }
 
+    protected long eatStartTime;
+    protected int totalEatTimeTicks;
     private void o() {
         if (this.isHandRaised()) {
             if (ItemStack.d(this.b(this.getRaisedHand()), this.activeItem)) {
@@ -2852,7 +2854,13 @@ public abstract class EntityLiving extends Entity {
                     this.b(this.activeItem, 5);
                 }
 
-                if (--this.bl == 0 && !this.world.isClientSide && !this.activeItem.m()) {
+                // Tuinity start - lag compensate eating
+                // we add 2 to the expected time to avoid lag compensating when we don't need to
+                boolean shouldLagCompensate
+                        = ca.spottedleaf.tuinity.config.TuinityConfig.lagCompensateEating && this.activeItem.getItem().isFood() && this.eatStartTime != -1 && (System.nanoTime() - this.eatStartTime) > ((2 + this.totalEatTimeTicks) * 50 * (1000 * 1000));
+                if ((--this.bl == 0 || shouldLagCompensate) && !this.world.isClientSide && !this.activeItem.m()) {
+                    this.setEatTimeTicks(0);
+                    // Tuinity end - lag compensate eating
                     this.q();
                 }
             } else {
@@ -2902,7 +2910,10 @@ public abstract class EntityLiving extends Entity {
 
         if (!itemstack.isEmpty() && !this.isHandRaised() || forceUpdate) { // Paper use override flag
             this.activeItem = itemstack;
-            this.bl = itemstack.k();
+            // Tuinity start - lag compensate eating
+            this.bl = this.totalEatTimeTicks = itemstack.k();
+            this.eatStartTime = System.nanoTime();
+            // Tuinity end - lag compensate eating
             if (!this.world.isClientSide) {
                 this.c(1, true);
                 this.c(2, enumhand == EnumHand.OFF_HAND);
@@ -2926,7 +2937,9 @@ public abstract class EntityLiving extends Entity {
                 }
             } else if (!this.isHandRaised() && !this.activeItem.isEmpty()) {
                 this.activeItem = ItemStack.a;
-                this.bl = 0;
+                this.bl = 0; // Tuinity - diff on change
+                this.totalEatTimeTicks = 0; // Tuinity - lag compensate eating
+                this.eatStartTime = -1L; // Tuinity - lag compensate eating
             }
         }
 
@@ -3048,7 +3061,9 @@ public abstract class EntityLiving extends Entity {
         }
 
         this.activeItem = ItemStack.a;
-        this.bl = 0;
+        this.bl = 0; // Tuinity - diff on change
+        this.totalEatTimeTicks = 0; // Tuinity - lag compensate eating
+        this.eatStartTime = -1L; // Tuinity - lag compensate eating
     }
 
     public boolean isBlocking() {
